### PR TITLE
Server prototype

### PR DESCRIPTION
**Pull request by KatKlo**

### CODE REVIEW

File: src/bsp_types.rs [hunk 1]
```diff
@@ -1,49 +1,37 @@
...
+            .map_err(Into::into)
     }
+}
 
-    fn parse_from_string(response_string: &'a str) -> Result<Self> {
-        let response: Response<Self> = serde_json::from_str(response_string)?;
-        Ok(response.result)
-    }
+pub trait MethodName {
+    fn get_method() -> &'static str;
```
Comment (PBundyra) [inline src/bsp_types.rs:26]: get_method_name() maybe?

File: src/server.rs [hunk 2]
```diff
@@ -1,53 +1,63 @@
...
+        if resp.success {
+            log(&format!(
+                "Received proper request from client: {:}\n",
+                &request_string
+            ));
+            log(&format!("Responding to client with: {:}\n", &resp.result));
+            send(&resp.result);
+        } else {
+            log(&format!(
+                "Received some string from client: {}\n response: {}\n",
```
Comment (PBundyra) [inline src/server.rs:44]: some -> invalid?

File: src/client.rs [hunk 3]
```diff
@@ -17,21 +23,41 @@ fn example_client_initialize_query() -> InitializeBuildParams {
...
 
-    for line in stdin().lock().lines() {
-        let line_string = line.unwrap();
+    pub fn run(&mut self) {
+        log("Client started\n");
 
-        if line_string.is_empty() {
-            break;
-        }
+        let request = example_client_initialize_query();
```
Comment (Toomimi) [inline src/client.rs:51]:
```suggestion
```

File: src/client.rs [hunk 4]
```diff
@@ -17,21 +23,41 @@ fn example_client_initialize_query() -> InitializeBuildParams {
...
+        let request_string = self
+            .create_request_string(RequestWrapper {
+                request_params: request,
+            })
+            .unwrap();
```
Comment (Toomimi) [inline src/client.rs:56]:
```suggestion
        let request_string = self
            .create_request_string(RequestWrapper {
                request_params: InitializeBuildParams::default(),
            })
            .unwrap();
```

File: src/client.rs [hunk 5]
```diff
@@ -1,7 +1,13 @@
...
+use jsonrpsee_core::traits::ToRpcParams;
+use jsonrpsee_types::{Id, RequestSer};
+use serde::Serialize;
+
+use crate::bsp_types::{
+    BuildClientCapabilities, InitializeBuildParams, MethodName, RequestWrapper,
+};
 use crate::utils::{log, send};
 
 fn example_client_initialize_query() -> InitializeBuildParams {
```
Comment (Toomimi) [inline src/client.rs:13]: This function is unused, can be deleted.

**Review by Toomimi — CHANGES_REQUESTED** [review]

**Review by PBundyra — COMMENTED** [review]
Looks good, GJ! However, src/client.rs and src/main.rs will require some changes during the merge with the redirect-io branch.

**Review by Toomimi — APPROVED** [review]

**Pull request merged**